src/dask_awkward/lib/optimize.py

```diff
@@ -86,177 +122,191 @@ def _projectable_io_layer_names(dsk: HighLevelGraph) -> list[str]:
-        )
-    pio_layer = pio_layer_names[0]
-
-    columns = dsk.layers[pio_layer]._meta.layout.form.columns()  # type: ignore
-
-    # can only select output columns that exist in the input
-    # (other names may have come from aliases)
-    keep = [c for c in out_meta.layout.form.columns() if c in columns]
-
-    for c in columns:
-        if c in keep:
-            continue
-        holdout = c
-        allcolumns = set(columns)
-        remaining = list(allcolumns - {holdout})
-        try:
-            _attempt_compute_with_columns_brute(dsk, columns=remaining)
-        except IndexError:
-            keep.append(holdout)
-    if keep == columns:
-        keep = None
-    return {pio_layer: keep}
+    reports = {}
+
+    # make labelled report
+    for name in _projectable_input_layer_names(dsk):
+        layers[name], report = layers[name].mock()
+        reports[name] = report
+
+    for name in _output_layer_names(dsk):
+        layers[name] = _mock_output(layers[name])
+
+    hlg = HighLevelGraph(layers, deps)
+    outlayer = list(hlg.layers.values())[-1]
+
+    try:
+        out = get_sync(hlg, list(outlayer.keys())[0])
+    except Exception as err:
+        on_fail = dask.config.get("awkward.optimization.on-fail")
+        # this is the default, throw a warning but skip the optimization.
+        if on_fail == "warn":
+            warnings.warn(f"Column projection optimization failed: {type(err)}, {err}")
+            return {}
+        # option "pass" means do not throw warning but skip the optimization.
+        elif on_fail == "pass":
+            log.debug("Column projection optimization failed; optimization skipped.")
+            return {}
+        # option "raise" to raise the exception here
+        elif on_fail == "raise":
+            raise
+        else:
+            raise ValueError(
+                f"Invalid awkward.optimization.on-fail option: {on_fail}.\n"
+                "Valid options are 'warn', 'pass', or 'raise'."
+            )
 
+    if isinstance(out, ak.Array):
+        out.layout._touch_data(recursive=True)
+    return reports
+
+
+def _necessary_columns(dsk: HighLevelGraph) -> dict[str, list[str]]:
+    """Pair layer names with lists of necessary columns."""
+    kv = {}
+    for name, report in _get_column_reports(dsk).items():
+        cols = {_ for _ in report.data_touched if _ is not None}
+        select = []
+        for col in sorted(cols):
+            if col == name:
+                continue
+            n, c = col.split(".", 1)
+            if n == name:
+                if c.endswith("__list__"):
+                    cnew = c[:-9].rstrip(".")
+                    if cnew not in select:
+                        select.append(f"{cnew}.*")
+                else:
+                    select.append(c)
+        kv[name] = select
+    return kv
+
+
+def _prune_wildcards(columns: list[str], meta: AwkwardArray) -> list[str]:
```

Comment:
@martindurant would be happy to get your take on this function.

It currently has one pitfall (that I know of!)

given a dataset with has parquet columns `foo.bar.x` and `foo.bar.y` (where `foo.bar` is a record array with fields `x` and `y`):

```python
dak.num(foo.bar.y, axis=1).compute()
```
will fail because `foo.bar.*` is what will be found as necessary, but then this function will result in `foo.bar.x` as the column to read (which is technically fine and should result in a correct computed array), but the dask graph will expect to get to `foo.bar.y`.

Reply:
> the dask graph will expect to get to foo.bar.y

What do you mean by this?

Reply:
This time it's the fully computed graph (not just the optimization typetracer graph) that errors. The optimization will determine that we only need _any_ field from `foo.bar` (which is technically true for the correct result), but the task graph has that additional getitem layer for retrieving `"y"`. 

I'm wondering if there's another piece of the graph we need to touch during the optimization to avoid dropping the field that was actually explicitly asked for in the original graph

Here's a minimal reproducer with the muons dataset, where `dak.num(ds.muons.eta, axis=1)` can be simplified to `dak.num(ds.muons, axis=1)` so we end up going down the path of just grabbing the first field of `ds.muons` (in this case it happens to be `pt`):

```python
>>> import dask_awkward as dak
>>> ds = dak.from_parquet("s3://ddavistemp/hpq", storage_options={"anon": True})
>>> dak.num(ds.muons.eta, axis=1)
dask.awkward<num, npartitions=50>
>>> a = dak.num(ds.muons.eta, axis=1)
>>> a.dask
HighLevelGraph with 4 layers.
<dask.highlevelgraph.HighLevelGraph object at 0x11e8afad0>
 0. read-parquet-1045eb08760cefeedd5f6194e5d1b18f
 1. muons-6a84d262864006fc9042dc8a4de45586
 2. eta-727d9bccb85d3046d4242f492b1a0c92
 3. num-76cb7a5f0a0c6ad8c1c7b3a45d97fa88

>>> a.compute()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/ddavis/software/repos/dask/dask/base.py", line 314, in compute
    (result,) = compute(self, traverse=False, **kwargs)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ddavis/software/repos/dask/dask/base.py", line 599, in compute
    results = schedule(dsk, keys, **kwargs)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ddavis/software/repos/dask/dask/threaded.py", line 89, in get
    results = get_async(
              ^^^^^^^^^^
  File "/Users/ddavis/software/repos/dask/dask/local.py", line 511, in get_async
    raise_exception(exc, tb)
  File "/Users/ddavis/software/repos/dask/dask/local.py", line 319, in reraise
    raise exc
  File "/Users/ddavis/software/repos/dask/dask/local.py", line 224, in execute_task
    result = _execute_task(task, data)
             ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ddavis/software/repos/dask/dask/core.py", line 119, in _execute_task
    return func(*(_execute_task(a, cache) for a in args))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ddavis/software/repos/dask/dask/optimization.py", line 990, in __call__
    return core.get(self.dsk, self.outkey, dict(zip(self.inkeys, args)))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ddavis/software/repos/dask/dask/core.py", line 149, in get
    result = _execute_task(task, cache)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ddavis/software/repos/dask/dask/core.py", line 119, in _execute_task
    return func(*(_execute_task(a, cache) for a in args))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ddavis/software/repos/dask/dask/core.py", line 119, in <genexpr>
    return func(*(_execute_task(a, cache) for a in args))
                  ^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ddavis/software/repos/dask/dask/core.py", line 113, in _execute_task
    return [_execute_task(a, cache) for a in arg]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ddavis/software/repos/dask/dask/core.py", line 113, in <listcomp>
    return [_execute_task(a, cache) for a in arg]
            ^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ddavis/software/repos/dask/dask/core.py", line 119, in _execute_task
    return func(*(_execute_task(a, cache) for a in args))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ddavis/.pyenv/versions/dev/lib/python3.11/site-packages/awkward/highlevel.py", line 951, in __getitem__
    out = self._layout[where]
          ~~~~~~~~~~~~^^^^^^^
  File "/Users/ddavis/.pyenv/versions/dev/lib/python3.11/site-packages/awkward/contents/content.py", line 514, in __getitem__
    return self._getitem(where)
           ^^^^^^^^^^^^^^^^^^^^
  File "/Users/ddavis/.pyenv/versions/dev/lib/python3.11/site-packages/awkward/contents/content.py", line 527, in _getitem
    return self._getitem_field(where)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ddavis/.pyenv/versions/dev/lib/python3.11/site-packages/awkward/contents/listoffsetarray.py", line 260, in _getitem_field
    self._content._getitem_field(where, only_fields),
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ddavis/.pyenv/versions/dev/lib/python3.11/site-packages/awkward/contents/unmaskedarray.py", line 197, in _getitem_field
    self._content._getitem_field(where, only_fields), parameters=None
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ddavis/.pyenv/versions/dev/lib/python3.11/site-packages/awkward/contents/recordarray.py", line 367, in _getitem_field
    return self.content(where)
           ^^^^^^^^^^^^^^^^^^^
  File "/Users/ddavis/.pyenv/versions/dev/lib/python3.11/site-packages/awkward/contents/recordarray.py", line 316, in content
    out = self.form_cls.content(self, index_or_field)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ddavis/.pyenv/versions/dev/lib/python3.11/site-packages/awkward/forms/recordform.py", line 153, in content
    index = self.field_to_index(index_or_field)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ddavis/.pyenv/versions/dev/lib/python3.11/site-packages/awkward/contents/recordarray.py", line 310, in field_to_index
    return self.form_cls.field_to_index(self, field)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ddavis/.pyenv/versions/dev/lib/python3.11/site-packages/awkward/forms/recordform.py", line 130, in field_to_index
    raise ak._errors.wrap_error(
awkward._errors.FieldNotFoundError: while attempting to slice

    <Array [[{pt: 43.1}, {...}, ..., {...}], ...] type='5994 * var * ?{pt: ...'>

with

    'eta'

Error details: no field 'eta' in record with 1 fields
```

Reply:
I see; that's ... annoying. So the metadata operation getting the field needs that field to exist even though we never touch its data. We could probably spot this pattern and apply an optimization transform, but that sounds like too much increased complexity.

Reply:
@jpivarski maybe you have an opinion here?

Reply:
I'm not hugely following this PR just now (lots of things in parallel), but as I see it we have two options:
1. Use the knowledge that the data aren't important, and use something like a NumPy strides trick to generate dummy data. This wouldn't work for for the JAX backend, but that's a fair tradeoff (JAX is limiting in places like this).
2. Don't apply this kind of optimisation. Not sure how feasible this is

Reply:
I'm not sure we can do the first. We need parquet or JSON to actually do some loading to get the offsets, but have no way to **only** get the offsets. We could modify the call in the real graph as to what we then do with it (e.g., override getitem to get what we loaded, whether it's the right field or not). Sounds a bit painful.

The column optimization should be pretty important. Yes, in this case we are able to load only the record, but it would be the full record to avoid the problem. That record could be all of the data actually. We might choose to make life easy and kick this problem to the future by selecting the whole record.

It is interesting that
```
dak.num(arr.record.field, axis=1) === dak.num(arr.record, axis=1)
```
but users probably don't realise this. The latter form means we can freely choose which column to load. The former might be the right one if the user knows which column should be the easiest to load.

Reply:
Some analysis objects can have quite a lot of columns, selecting the whole record only to count it would be extremely expensive in terms of bandwidth off disk (especially for jets, for instance).

Reply:
I think right now our best bet is to perform the optimization as-is and be prepared to field issues/questions based on the error message. We can also try to come up with a decent piece of docs explaining this potential obstacle